src/Mvc/Mvc.Core/src/ControllerBase.cs

```diff
@@ -1821,6 +1822,52 @@ public virtual ConflictObjectResult Conflict([ActionResultObjectValue] object er
         public virtual ConflictObjectResult Conflict([ActionResultObjectValue] ModelStateDictionary modelState)
             => new ConflictObjectResult(modelState);
 
+        /// <summary>
+        /// Creates an <see cref="ObjectResult"/> that produces a <see cref="ProblemDetails"/> response with a <c>500</c>
+        /// error status with a <see cref="ProblemDetails" /> value.
+        /// </summary>
+        /// <param name="title">The value for <see cref="ProblemDetails.Title" />.</param>
+        /// <param name="type">The value for <see cref="ProblemDetails.Type" />.</param>
+        /// <param name="detail">The value for <see cref="ProblemDetails.Detail" />.</param>
+        /// <param name="instance">The value for <see cref="ProblemDetails.Instance" />.</param>
+        /// <returns>The created <see cref="ObjectResult"/> for the response.</returns>
+        [NonAction]
+        public virtual ObjectResult Problem(
+            string detail = null,
+            string instance = null,
+            string title = null,
+            string type = null)
+        {
+            var problemDetails = new ProblemDetails
+            {
+                Title = title,
+                Type = type,
+                Detail = detail,
+                Instance = instance,
+            };
+
+            ApplyProblemDetailsDefaults(problemDetails, statusCode: 500);
+
+            return new ObjectResult(problemDetails);
+        }
+
+        private void ApplyProblemDetailsDefaults(ProblemDetails problemDetails, int statusCode)
+        {
+            problemDetails.Status = statusCode;
+
+            if (problemDetails.Title is null || problemDetails.Type is null)
+            {
+                var options = HttpContext.RequestServices.GetRequiredService<IOptions<ApiBehaviorOptions>>().Value;
```

Comment:
I'm a little anxious about putting a service here without a way to mock it. I don't have a great solution other than doing things like making this a filter.

Most of the other places we interact with services, you can mock them easily. https://github.com/aspnet/AspNetCore/blob/f7e3eaa7e324b99aeb496a68414cd093acaae802/src/Mvc/Mvc.Core/src/ControllerBase.cs#L171

Does it seem reasonable to make a new service, and then add that as a property, and then call it from here?

Reply:
Sure.